### PR TITLE
fix #804 - add a parameter to disable swagger ui when not needed

### LINKED ISSUE #804: Swagger ui overwrites html encoder response
I have recently updated `api-platform/core` from `beta` to `rc2`. Since that the swagger ui overwrites all our html encoder responses. I feel like there is something wrong with the condition here: https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/Bundle/EventListener/SwaggerUiListener.php#L27-L28

If we request for instance an url like `http://fancy-api.com/documents/1` and we have an html encoder registered the condition for exiting that function are not met:

``` php
if ('html' !== $request->getRequestFormat(null) 
    || (!$request->attributes->has('_api_resource_class') && !$request->attributes->has('_api_respond'))
    ) {
        return;
    }
```

reads:

``` php
if (false 
    || (!true && !false)
    ) {
        return;
    }
```

I think in the earlier version the swagger ui was only triggered on the route `/apidoc`. Maybe it would be possible to have something like this:

`/doc` --> swagger ui
`/doc.jsonld` --> hydra doc

What do you guys think?

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #804 
| License       | MIT
| Doc PR        | O

This PR add a parameter to disable swagger ui

### CODE REVIEW

**Comment by soyuka** [comment]
+1 this can be needed in correlation with #768

**Comment by dunglas** [comment]
I suggest to don't register `SwaggerUiListener` at all if it Swagger UI deactivated instead of passing a flag in the listener constructor. Cleaner code, and better performance.

**Comment by Simperfit** [comment]
@dunglas I've made the change

File: src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php [hunk 1]
```diff
@@ -141,7 +141,13 @@ private function registerSwaggerConfiguration(ContainerBuilder $container, array
         }
 
         $loader->load('swagger.xml');
-        $container->setParameter('api_platform.enable_swagger', (string) $config['enable_swagger']);
+
+        if ($config['enable_swagger_ui']) {
+            $loader->load('swagger-ui.xml');
+            $container->setParameter('api_platform.enable_swagger_ui', $config['enable_swagger_ui']);
```
Comment (meyerbaptiste) [inline src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php]: And if `$config['enable_swagger_ui']` is false?

Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php]: It does not load swagger-ui, I don't understand what you mean

Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php]: OH: We don't net to set it for the container if it's false anyway. Like we have done with enable_swagger.

**Comment by dunglas** [comment]
Thank you @Simperfit

**Comment by teohhanhui** [comment]
This does not address the use case of having Swagger UI, but disabling the "hijacking".

**Comment by dunglas** [comment]
@teohhanhui yes as long as you register yourself the Swagger UI controller: https://github.com/api-platform/core/issues/804#issuecomment-259955376

**Comment by teohhanhui** [comment]
That's not very nice for DX :disappointed:

**Comment by Simperfit** [comment]
@teohhanhui I think it's okay, not alot of people use the html encoders and if they do we will have to add a doc part about this.

What would be better ?

**Comment by teohhanhui** [comment]
Not many people use a HTML encoder, yes. But many people like myself would not want the hijacking (mostly just a marketing feature to me).

It'd be much better if the hijacking can be disabled by configuration.

I propose something like the following:

* `SwaggerUiListener` should only handle API Platform paths (focus on the hijacking), not `/docs` etc.
* `/docs` should be registered explicitly in the app's routing configuration, the way it used to be.

**Comment by soyuka** [comment]
I made a swagger configuration node in #768 :

```yaml
swagger:
  enabled: true
  ui_enabled: true
  ui_path: '%kernel.root_dir%/../web/swagger-ui
```

I'd propose:


```yaml
swagger:
  enabled: true
  ui:
    enabled: true
    path: '%kernel.root_dir%/../web/swagger-ui
    intercept_operations: true 
```

**Pull request merged**